src/licenses/LicenseModel.cpp

```diff
@@ -0,0 +1,184 @@
+
+    while (it.hasNext()) {
+      QString path = it.next();
+
+      QList<QVariant> categories;
+      categories.append(QVariant(static_cast<int>(LicenseCategory::All)));
+
+      if (path.startsWith(webviewRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::WebView)));
+
+      if (path.startsWith(toolkitRootPath) && !path.startsWith(webviewRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::Toolkit)));
+
+      if (path.startsWith(platformRootPath) && !path.startsWith(qtRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::Platform)));
+
+      const QFileInfo &fileInfo = it.fileInfo();
+      LicenseItem *parent = map[fileInfo.dir().path()];
+
+      QVector<QVariant> child_data = {fileInfo.fileName(), fileInfo.absoluteFilePath()};
+      auto *child = new LicenseItem(child_data, categories, parent);
+      parent->appendChild(child);
+
+      if (fileInfo.isDir())
+        map[child->path()] = child;
+    }
+  }
+
+  QModelIndex LicenseModel::index(int row, int column, const QModelIndex &parent) const {
+    if (!hasIndex(row, column, parent))
+      return QModelIndex();
+
+    LicenseItem *parentItem;
+
+    if (!parent.isValid())
+      parentItem = rootItem.get();
+    else
+      parentItem = static_cast<LicenseItem *>(parent.internalPointer());
+
+    LicenseItem *childItem = parentItem->child(row);
+    if (childItem)
+      return createIndex(row, column, childItem);
+    return QModelIndex();
+  }
+
+  QModelIndex LicenseModel::parent(const QModelIndex &index) const {
+    if (!index.isValid())
+      return QModelIndex();
+
+    auto *childItem = static_cast<LicenseItem *>(index.internalPointer());
+    LicenseItem *parentItem = childItem->parentItem();
+
+    if (parentItem == rootItem.get())
+      return QModelIndex();
+
+    return createIndex(parentItem->row(), 0, parentItem);
+  }
+
+  int LicenseModel::rowCount(const QModelIndex &parent) const {
+    LicenseItem *parentItem;
+    if (parent.column() > 0)
+      return 0;
+
+    if (!parent.isValid())
+      parentItem = rootItem.get();
+    else
+      parentItem = static_cast<LicenseItem *>(parent.internalPointer());
+
+    return parentItem->childCount();
+  }
+
+  int LicenseModel::columnCount(const QModelIndex &parent) const {
+    if (parent.isValid())
+      return static_cast<LicenseItem *>(parent.internalPointer())->columnCount();
+    return rootItem->columnCount();
+  }
+
+  QVariant LicenseModel::data(const QModelIndex &index, int role) const {
+    if (!index.isValid())
+      return QVariant();
+
+    if (role != licenses::LicenseFileName && role != licenses::LicenseFilePath && role != licenses::LicenseCategories)
+      return QVariant();
+
+    auto *item = static_cast<LicenseItem *>(index.internalPointer());
+
+    return item->data(role);
+  }
+
+  Qt::ItemFlags LicenseModel::flags(const QModelIndex &index) const {
+    if (!index.isValid())
```

Comment:
what does QAbstractItemModel::flags(index) do when index is invalid?  Does it do the same as this?

When is index invalid anyhow?  Does that happen or are you just being extra safe?

Reply:
I removed the whole function - I don't believe overriding it is necessary atm